### ISSUE TITLE
Test suite hangs on test_that_threads_are_different

### ISSUE THREAD

**Issue by jMyles** [hendrix/hendrix#140]
It's not clear why this started during the process of adding compatibility for Python 3 / Twisted 17, but here's a conversation from the #twisted IRC channel:

```
<runciter> jMyles: so i can tell you that the issue is: blockingCallFromThread is being run in the same thread as the reactor
<runciter> so you have a deadlock
<runciter> i'm still tracing through why that's the case
<jMyles> runciter: Yeah, I mean, I figured I must be blocking the reactor somewhere
<runciter> (you can prove this by putting 
<runciter> ugh, h and j are so close
<runciter> jMyles: the callable might run but the result won't come back until the reactor's started
<runciter> since it relies on callFromThread
<runciter> jMyles: i think the category of problem is that WSGIResource.whatever is supposed to be called from the main thread, and you're calling it from within a threadpool
<runciter> i'm not sure yet, though
<runciter> i'm trying to multitask this in
<jMyles> Yeah?  I have always called render() in a thread - isn't that how Twisted Web is?
<jMyles> The parent class also returns NOT_DONE_YET (ie, not another Deferred).  This made me think it was cool to call in a thread.
<runciter> jMyles: ok, so attaching to the process with gdb and running pystack shows that it's blocked on ThreadPool.stop
<runciter> jMyles: oh, dear
<jMyles> runciter: Interesting.  I tried pressing the "pause" button in PyCharm, which normally has the same effect, and it did nothing.  So, I learned something today.
<runciter> pycharm makes some... interesting assumptions about how you're running python IME
<jMyles> I was lukewarm about it 3 years ago, and since then it seems to have steadily added and improved all the things I care about.  Now it's my favorite editing tool.
<jMyles> runciter: So, if I return None instead of NOT_DONE_YET, it does not hang.
<runciter> jMyles: there seems to be a race condition
<runciter> jMyles: because sprinkling print calls and stepping through with pdb sometimes causes it to pass
<runciter> jMyles: so the deferred returned from request_same _or_different_thread_thread seems to reliably fire with NOT_DONE_YET
<jMyles> runciter: Yeah, must be a race condition.  My queues must have come unscrewed.  :-)
<runciter> jMyles: also there's `lambda _: self.assert_that_threads_are_different` when it should be `lambda _: self.assert_that_threads_are_different()`
<runciter> that's not actually related
<runciter> jMyles: so i think deadlock is approximately: HendrixWSGIResource.write creates a HendrixWSGIResponse and starts it, then returns NOT_DONE_YET
<runciter> jMyles: now the main reactor thread resolves the deferred from request_same_or_different_thread_thread, which fires with NOT_DONE_YET
<runciter> jMyles: and it successfully runs its callbacks (lambda _: ...), finishes the test, and then attempts to run its cleanup functions
<runciter> jMyles: now it's blocked on tp.stop()
<runciter> jMyles: going back to HendrixWSGIResponse.start() in HendrixWSGIResource.render, a thread in the threadpool starts running _WSGIResource.run
<runciter> jMyles: eventually this wants to write to the transport, which does blockingCallFromThread
<runciter> but this happens _after_ the main thread is blocked on tp.stop()
<runciter> so there's a thread in self.tp that's waiting for the reactor in the main thread to tell it that its write has finished, but the main thread is blocked on tp.stop()
<runciter> and tp.stop() can't return because the thread in self.tp is blocked waiting for the reactor...
<runciter> thus a deadlock
<runciter> so i think a solution is to make the test wait for request.notifyFinish()
<runciter> let's see
<runciter> yes, that fixes it
<runciter> jMyles: add `d.addCallback(lambda _: request1.notifyFinish())` after `d = deferToThreadPool(reactor, self.tp, hr.render, request1)` and before `return d` in `request_same_or_different_thread_thread` 
<runciter> so this way, the reactor thread doesn't reach tp.stop() until *after* the worker thread's write (via blockingCallFromThread) completes
```
